File: extension/neovim/lua/fluidcad/code_api.lua
local M = {}

local function get_url()
  local ok, bridge = pcall(require, 'fluidcad.bridge')
  if not ok then
    return nil
  end
  return bridge.get_url()
end

--- POST a JSON body to /api/code/<endpoint> and return the decoded response,
--- or nil on any failure (server unreachable, non-2xx, decode error).
local function post(endpoint, body)
  local url = get_url()
  if not url or url == '' then
    return nil
  end
  local encoded = vim.fn.json_encode(body)
  local result = vim.fn.system(
    { 'curl', '-s', '-f', '-X', 'POST',
      '-H', 'Content-Type: application/json',
      '--data-binary', '@-',
      url .. '/api/code/' .. endpoint },
    encoded
  )
  if vim.v.shell_error ~= 0 then
    return nil
  end
  local ok, decoded = pcall(vim.fn.json_decode, result)
  if not ok or type(decoded) ~= 'table' then
    return nil
  end
  return decoded
end

function M.add_breakpoint(code, reference_row)
  return post('add-breakpoint', { code = code, referenceRow = reference_row })
end

function M.remove_breakpoint(code, line)
  return post('remove-breakpoint', { code = code, line = line })
end

function M.toggle_breakpoint(code, cursor_row)
  return post('toggle-breakpoint', { code = code, cursorRow = cursor_row })
end

function M.clear_breakpoints(code)
  return post('clear-breakpoints', { code = code })
end

function M.insert_point(code, source_line, point)
  return post('insert-point', { code = code, sourceLine = source_line, point = point })
end

function M.remove_point(code, source_line, point)
  return post('remove-point', { code = code, sourceLine = source_line, point = point })
end

function M.add_pick(code, source_line)
  return post('add-pick', { code = code, sourceLine = source_line })
end

function M.remove_pick(code, source_line)
  return post('remove-pick', { code = code, sourceLine = source_line })
end

function M.set_pick_points(code, source_line, points)
  return post('set-pick-points', { code = code, sourceLine = source_line, points = points })
end

function M.insert_geometry(code, sketch_source_line, statement, new_variable)
  return post('insert-geometry', {
    code = code,
    sketchSourceLine = sketch_source_line,
    statement = statement,
    newVariable = new_variable,
  })
end

function M.update_position(code, source_line, new_position, point_index)
  return post('update-position', { code = code, sourceLine = source_line, newPosition = new_position, pointIndex = point_index or 0 })
end

function M.set_line_position(code, source_line, new_start, new_end)
  return post('set-line-position', { code = code, sourceLine = source_line, newStart = new_start, newEnd = new_end })
end

function M.set_chain_positions(code, source_line, updates)
  return post('set-chain-positions', { code = code, sourceLine = source_line, updates = updates })
end

function M.set_rect_dimensions(code, source_line, start_point, width, height)
  return post('set-rect-dimensions', { code = code, sourceLine = source_line, startPoint = start_point, width = width, height = height })
end

function M.update_dimension(code, source_line, new_value)
  return post('update-dimension', { code = code, sourceLine = source_line, newValue = new_value })
end

function M.update_dimension_expression(code, source_line, expression, sketch_source_line, new_variable, dimension_offset)
  return post('update-dimension-expression', {
    code = code,
    sourceLine = source_line,
    expression = expression,
    sketchSourceLine = sketch_source_line,
    newVariable = new_variable,
    dimensionOffset = dimension_offset or 0,
  })
end

--- Replace the entire contents of `bufnr` with `new_code`. Returns true on
--- success.
function M.replace_buffer(bufnr, new_code)
  if not vim.api.nvim_buf_is_valid(bufnr) then
    return false
  end
  local lines = vim.split(new_code, '\n', { plain = true })
  vim.api.nvim_buf_set_lines(bufnr, 0, -1, false, lines)
  return true
end

return M

File: extension/vscode/src/code-api.ts
import * as vscode from 'vscode';

export type BreakpointEditResult = { newCode: string; breakpointLine: number | null };
export type CodeEditResult = { newCode: string };

async function postCodeEdit<T>(
  serverUrl: string,
  endpoint: string,
  body: Record<string, unknown>,
  logger: vscode.OutputChannel,
): Promise<T | null> {
  if (!serverUrl) {
    return null;
  }
  try {
    const response = await fetch(`${serverUrl}/api/code/${endpoint}`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
    });
    if (!response.ok) {
      logger.appendLine(`[code-api] ${endpoint} returned HTTP ${response.status}`);
      return null;
    }
    return (await response.json()) as T;
  } catch (err) {
    logger.appendLine(`[code-api] ${endpoint} failed: ${err}`);
    return null;
  }
}

export function addBreakpoint(serverUrl: string, code: string, referenceRow: number, logger: vscode.OutputChannel) {
  return postCodeEdit<BreakpointEditResult>(serverUrl, 'add-breakpoint', { code, referenceRow }, logger);
}

export function removeBreakpoint(serverUrl: string, code: string, line: number, logger: vscode.OutputChannel) {
  return postCodeEdit<BreakpointEditResult>(serverUrl, 'remove-breakpoint', { code, line }, logger);
}

export function toggleBreakpoint(serverUrl: string, code: string, cursorRow: number, logger: vscode.OutputChannel) {
  return postCodeEdit<BreakpointEditResult>(serverUrl, 'toggle-breakpoint', { code, cursorRow }, logger);
}

export function clearBreakpoints(serverUrl: string, code: string, logger: vscode.OutputChannel) {
  return postCodeEdit<CodeEditResult>(serverUrl, 'clear-breakpoints', { code }, logger);
}

export function insertPoint(
  serverUrl: string, code: string, sourceLine: number, point: [number, number], logger: vscode.OutputChannel,
) {
  return postCodeEdit<CodeEditResult>(serverUrl, 'insert-point', { code, sourceLine, point }, logger);
}

export function removePoint(
  serverUrl: string, code: string, sourceLine: number, point: [number, number], logger: vscode.OutputChannel,
) {
  return postCodeEdit<CodeEditResult>(serverUrl, 'remove-point', { code, sourceLine, point }, logger);
}

export function addPick(serverUrl: string, code: string, sourceLine: number, logger: vscode.OutputChannel) {
  return postCodeEdit<CodeEditResult>(serverUrl, 'add-pick', { code, sourceLine }, logger);
}

export function removePick(serverUrl: string, code: string, sourceLine: number, logger: vscode.OutputChannel) {
  return postCodeEdit<CodeEditResult>(serverUrl, 'remove-pick', { code, sourceLine }, logger);
}

export function addGuide(serverUrl: string, code: string, sourceLine: number, logger: vscode.OutputChannel) {
  return postCodeEdit<CodeEditResult>(serverUrl, 'add-guide', { code, sourceLine }, logger);
}

export function removeGuide(serverUrl: string, code: string, sourceLine: number, logger: vscode.OutputChannel) {
  return postCodeEdit<CodeEditResult>(serverUrl, 'remove-guide', { code, sourceLine }, logger);
}

export function removeStatement(serverUrl: string, code: string, sourceLine: number, logger: vscode.OutputChannel) {
  return postCodeEdit<CodeEditResult>(serverUrl, 'remove-statement', { code, sourceLine }, logger);
}

export function setFeatureName(
  serverUrl: string, code: string, sourceLine: number, name: string | null, logger: vscode.OutputChannel,
) {
  return postCodeEdit<CodeEditResult>(serverUrl, 'set-feature-name', { code, sourceLine, name }, logger);
}

export function insertLoad(serverUrl: string, code: string, fileName: string, logger: vscode.OutputChannel) {
  return postCodeEdit<CodeEditResult>(serverUrl, 'insert-load', { code, fileName }, logger);
}

export function setPickPoints(
  serverUrl: string, code: string, sourceLine: number, points: [number, number][], logger: vscode.OutputChannel,
) {
  return postCodeEdit<CodeEditResult>(serverUrl, 'set-pick-points', { code, sourceLine, points }, logger);
}

export type InsertChainEdit = {
  ground?: boolean;
  name?: string | null;
  defaultName?: string;
  translate?: [number, number, number] | null;
};

export function updateInsertChain(
  serverUrl: string,
  code: string,
  sourceLine: number,
  edit: InsertChainEdit,
  logger: vscode.OutputChannel,
) {
  return postCodeEdit<CodeEditResult>(
    serverUrl, 'update-insert-chain', { code, sourceLine, edit }, logger,
  );
}

export function insertGeometry(
  serverUrl: string, code: string, sketchSourceLine: number, statement: string, logger: vscode.OutputChannel,
  newVariable: { name: string; initializer: string } | null = null,
) {
  return postCodeEdit<CodeEditResult>(
    serverUrl, 'insert-geometry',
    { code, sketchSourceLine, statement, newVariable },
    logger,
  );
}


export type SketchPositionEditPayload = {
  sourceLine: number;
  points?: { pointIndex: number; position: [number, number]; expected?: [number, number] }[];
  scalar?: { value: number; expected?: number };
};

/** Unlike the other transforms this one can refuse (drift guard) — the
 * refusal rides `error` beside the untouched code. */
export type SketchPositionsResult = { newCode: string; error?: string };

export function updateSketchPositions(
  serverUrl: string, code: string,
  edits: SketchPositionEditPayload[], logger: vscode.OutputChannel,
) {
  return postCodeEdit<SketchPositionsResult>(serverUrl, 'update-sketch-positions', { code, edits }, logger);
}

export function updateDimension(
  serverUrl: string, code: string, sourceLine: number, newValue: number, logger: vscode.OutputChannel,
) {
  return postCodeEdit<CodeEditResult>(serverUrl, 'update-dimension', { code, sourceLine, newValue }, logger);
}

export function updateDimensionExpression(
  serverUrl: string, code: string, sourceLine: number, expression: string, logger: vscode.OutputChannel,
  sketchSourceLine: number | null = null,
  newVariable: { name: string; initializer: string } | null = null,
  dimensionOffset = 0,
) {
  return postCodeEdit<CodeEditResult>(
    serverUrl, 'update-dimension-expression',
    { code, sourceLine, expression, sketchSourceLine, newVariable, dimensionOffset },
    logger,
  );
}

/**
 * `filePath` lets the server refuse an assembly file (422 → null here). A
 * null `unit` removes the file's declaration ("Same as project") — it is
 * sent as JSON null, which the route accepts, never dropped.
 */
export function setUnit(
  serverUrl: string, code: string, unit: string | null, filePath: string, logger: vscode.OutputChannel,
) {
  return postCodeEdit<CodeEditResult>(serverUrl, 'set-unit', { code, unit, filePath }, logger);
}

export function applyFeature(
  serverUrl: string, code: string, spec: unknown, logger: vscode.OutputChannel,
) {
  return postCodeEdit<CodeEditResult & { error?: string }>(serverUrl, 'apply-feature', { code, spec }, logger);
}


/**
 * Replace the entire contents of `doc` with `newCode` in a single workspace
 * edit. Returns true on success.
 */
export async function replaceDocument(doc: vscode.TextDocument, newCode: string): Promise<boolean> {
  if (doc.getText() === newCode) {
    return true;
  }
  const fullRange = new vscode.Range(
    new vscode.Position(0, 0),
    doc.lineAt(doc.lineCount - 1).range.end,
  );
  const edit = new vscode.WorkspaceEdit();
  edit.replace(doc.uri, fullRange, newCode);
  return vscode.workspace.applyEdit(edit);
}
